### PR TITLE
Adding JLBP-14: Do not use version ranges

### PR DESCRIPTION
**Pull request by garrettjonesgoogle**

### CODE REVIEW

File: library-best-practices/JLBP-14.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+[JLBP-14] Use explicit versions, not version ranges
+---------------------------------------------------
+
+- Neither Maven nor Gradle have have a mechanism to locally lock versions like
+  package managers in other languages. If the pom.xml of your library specifies
+  a version range instead of an explicit version, it can pull in snapshot
```
Comment (elharo) [inline library-best-practices/JLBP-14.md]: I'm not sure it will get a snapshot version in this case.

File: library-best-practices/JLBP-14.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
+[JLBP-14] Use explicit versions, not version ranges
+---------------------------------------------------
+
+- Neither Maven nor Gradle have have a mechanism to locally lock versions like
```
Comment (elharo) [inline library-best-practices/JLBP-14.md]: Is the first sentence needed?

File: library-best-practices/JLBP-14.md (hunk 1 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-14.md]: Ok I wasn't sure myself - I think this statement was written by someone else and rode along to my PR. I'll just remove the reference to snapshots.

File: library-best-practices/JLBP-14.md (hunk 2 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-14.md]: No it's honestly not needed, since the JLBPs are focused on Java library authors. I will remove it.

**Comment by garrettjonesgoogle** [comment]
Feedback addressed, PTAL

File: library-best-practices/JLBP-14.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
+[JLBP-14] Use explicit versions, not version ranges
+---------------------------------------------------
+
+- If the pom.xml of your library specifies a version range instead of an
+  explicit version, builds at different points in time can get different
+  dependency versions that could break your library unexpectedly.
+  - Version ranges can even pull in incomplete releases that will your
+    library's build (example:
+    https://github.com/GoogleCloudPlatform/appengine-gcs-client/issues/71).
+  - Thus, always use an explicit version.
```
[the remark this answers is not in the document]
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-14.md]: You make a good point. I'll change references to "single" version.

File: library-best-practices/JLBP-14.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+- Single-element version ranges ("hard requirements") have a much different
```
[the remark this answers is not in the document]
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-14.md]: The "standard" way of specifying a version using a single version number is a "soft" requirement (it allows maven to select a higher version depending on its resolution rules). The "hard" requirement disallows anything but the exact version.

https://maven.apache.org/pom.html#Dependency_Version_Requirement_Specification

- 1.0: "Soft" requirement on 1.0 (just a recommendation, if it matches all other ranges for the dependency)
- [1.0]: "Hard" requirement on 1.0

File: library-best-practices/JLBP-14.md [hunk 5]
```diff
@@ -0,0 +1,12 @@
+[JLBP-14] Use explicit versions, not version ranges
+---------------------------------------------------
+
+- If the pom.xml of your library specifies a version range instead of an
+  explicit version, builds at different points in time can get different
+  dependency versions that could break your library unexpectedly.
+  - Version ranges can even pull in incomplete releases that will your
```
[the remark this answers is not in the document]
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-14.md]: oops, will fix

File: library-best-practices/JLBP-14.md [hunk 6]
```diff
@@ -0,0 +1,12 @@
+[JLBP-14] Use explicit versions, not version ranges
+---------------------------------------------------
+
+- If the pom.xml of your library specifies a version range instead of an
+  explicit version, builds at different points in time can get different
```
[the remark this answers is not in the document]
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-14.md]: will fix

**Comment by garrettjonesgoogle** [comment]
Feedback addressed, PTAL

File: library-best-practices/JLBP-14.md [hunk 7]
```diff
@@ -0,0 +1,16 @@
+[JLBP-14] Do not use version ranges
+-----------------------------------
+
+- If the pom.xml of your library specifies a version range instead of a
+  single version for any particular dependency, builds at different points
+  in time can get different versions of that dependency (as your dependency
```
Comment (elharo) [inline library-best-practices/JLBP-14.md]: perhaps, can get --> see

File: library-best-practices/JLBP-14.md [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+  releases new versions), which could break your library unexpectedly.
```
Comment (elharo) [inline library-best-practices/JLBP-14.md]: could --> can

File: library-best-practices/JLBP-14.md [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+  - Version ranges can even pull in incomplete releases that break your
+    library's build (example:
```
Comment (elharo) [inline library-best-practices/JLBP-14.md]: Random thought: did this qualify as a static linkage error or it a different kind of problem?

File: library-best-practices/JLBP-14.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (elharo) [reply library-best-practices/JLBP-14.md]: Weird. I never heard of anything like that before. It doesn't seem to explain what a hard or soft requirement is. We should probably double check with maven and make sure that isn't some legacy detail that isn't true any more.

[the remark this answers is not in the document]
Comment (elharo) [reply library-best-practices/JLBP-14.md]: https://books.sonatype.com/mvnref-book/reference/pom-relationships-sect-project-dependencies.html#pom-relationships-sect-version-ranges

This appears to be very uncommonly used. I wonder if we should recommend against hard requirements?

File: library-best-practices/JLBP-14.md (hunk 7 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-14.md]: I think the "can" part is still necessary; if a dependency never actually updates, you won't see different versions. I will s/get/see/ though.

File: library-best-practices/JLBP-14.md (hunk 8 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-14.md]: Ok

File: library-best-practices/JLBP-14.md (hunk 9 above)
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-14.md]: I would say this is a different type of problem; the dependency tree that was specified couldn't be fetched.

File: library-best-practices/JLBP-14.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (garrettjonesgoogle) [reply library-best-practices/JLBP-14.md]: grpc-java is currently using it. I don't think it's a bad thing; it is a way to enforce something like the fact that grpc-auth 1.15.0 must be used with grpc-core 1.15.0.

**Pull request merged**